### PR TITLE
Add thermal deformation gradient for CP

### PR DESCRIPTION
**Pull request by dewenyushu**
Add thermal induced deformation in the material-based crystal plasticity base class.

Refs #17734

### CODE REVIEW

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C [hunk 1]
```diff
@@ -77,6 +82,8 @@ ComputeMultipleCrystalPlasticityStress::ComputeMultipleCrystalPlasticityStress(
     _line_search_method(getParam<MooseEnum>("line_search_method").getEnum<LineSearchMethod>()),
     _plastic_deformation_gradient(declareProperty<RankTwoTensor>("fp")),
     _plastic_deformation_gradient_old(getMaterialPropertyOld<RankTwoTensor>("fp")),
+    _thermal_deformation_gradient(declareProperty<RankTwoTensor>("fth")),
+    _thermal_deformation_gradient_old(getMaterialPropertyOld<RankTwoTensor>("fth")),
```
Comment (jiangwen84) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C:86]: Shall we  declare `thermal_deformation_gradient` only when it is coupled to temperature?

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C [hunk 2]
```diff
@@ -85,8 +92,14 @@ ComputeMultipleCrystalPlasticityStress::ComputeMultipleCrystalPlasticityStress(
...
     _crysrot(getMaterialProperty<RankTwoTensor>(
-        "crysrot")) // defined in the elasticity tensor classes for crystal plasticity
+        "crysrot")), // defined in the elasticity tensor classes for crystal plasticity
+    _temperature(isParamValid("temperature") ? coupledValue("temperature") : _zero),
+    _temperature_old(isParamValid("temperature") ? coupledValueOld("temperature") : _zero)
 {
+  if (isParamValid("temperature"))
+    _thermal_expansion_coefficients.fillFromInputVector(
+        getParam<std::vector<Real>>("thermal_expansion_coefficients"));
+
```
Comment (jiangwen84) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C:102]: maybe you want to check the vector size?

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C [hunk 3]
```diff
@@ -77,6 +82,8 @@ ComputeMultipleCrystalPlasticityStress::ComputeMultipleCrystalPlasticityStress(
     _line_search_method(getParam<MooseEnum>("line_search_method").getEnum<LineSearchMethod>()),
     _plastic_deformation_gradient(declareProperty<RankTwoTensor>("fp")),
     _plastic_deformation_gradient_old(getMaterialPropertyOld<RankTwoTensor>("fp")),
```
Comment (jiangwen84) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C:84]: Can we switch to full names for those material properties? 

`fp` --> `plastic_deformation_gradient`
`pk2` --> `second_piola_kirchhoff_stress`
...

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C [hunk 4]
```diff
@@ -446,10 +471,19 @@ ComputeMultipleCrystalPlasticityStress::calculateResidual()
...
+  // calculate F^{th -1}
+  Real dtheta = (_temperature[_qp] - _temperature_old[_qp]) * _substep_dt / _dt;
+  RankTwoTensor residual_equivalent_thermal_expansion_increment =
+      RankTwoTensor::Identity() - dtheta * _thermal_expansion_coefficients;
+  _inverse_thermal_deformation_grad =
+      _inverse_thermal_deformation_grad_old * residual_equivalent_thermal_expansion_increment;
+
+  _elastic_deformation_gradient = _temporary_deformation_gradient *
+                                  _inverse_thermal_deformation_grad *
+                                  _inverse_plastic_deformation_grad;
```
Comment (jiangwen84) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C:483]: Can we avoid those calculation if users do not use thermal eigenstrain?

File: modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C [hunk 5]
```diff
@@ -329,25 +329,37 @@ CrystalPlasticityStressUpdateBase::calculateSchmidTensor(
...
+    _tau[_qp][i] = pk2_hat.doubleContraction(_flow_direction[_qp][i]);
+  }
 }
 
 void
 CrystalPlasticityStressUpdateBase::calculateTotalPlasticDeformationGradientDerivative(
-    RankFourTensor & dfpinvdpk2, const RankTwoTensor & inverse_plastic_deformation_grad_old)
+    RankFourTensor & dfpinvdpk2,
+    const RankTwoTensor & inverse_plastic_deformation_grad_old,
+    const RankTwoTensor & inverse_thermal_deformation_grad_old)
```
Comment (jiangwen84) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C:349]: Those codes seem to be hard to extend to other eigenstrains, e.g., irradiation eigenstrain. Do you have any ideas how to include genetic eigenstrains? This is not required for this PR, but it will be nice to have some thought here.

**Review by jiangwen84 — CHANGES_REQUESTED** [review]
Looks great! Some minor comments.

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C (hunk 1 above)
Comment (dewenyushu) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C:86]: Thanks @jiangwen84 . Good point. I will make this change.

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C (hunk 3 above)
Comment (dewenyushu) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C:84]: Sure, I can make this change.

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C (hunk 4 above)
Comment (dewenyushu) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C:483]: Fth is identity if temperature is a constant so I was assuming no/not much computation is added here. But yeah, I agree it is better to avoid the computation if no thermal coupling is specified

File: modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C (hunk 5 above)
Comment (dewenyushu) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C:349]: Glad that you asked! I think possible to make it more flexible in terms of additional eigenstrain types. I have to re-derive some equations further in terms of how this affects the Jacobian calculations. But from what I can see so far, it is possible to have one deformation gradient, e.g., F_eigen = F_th * F_irradiation * ..., in the base class, which encompasses all the additional eigenstrain contributions.  

I will definitely need to discuss with you more once I start extending this capability.

**Comment by dewenyushu** [comment]
@jiangwen84 would you mind taking another look (especially at the part where I am trying to avoid calculations when thermal coupling is not needed) and let me know if there are some further issues? Really appreciate it!

**Comment by hugary1995** [comment]
@dewenyushu @jiangwen84 I have a general suggestion regarding the design:

There is a more general definition of thermal deformation gradient, which is eigen deformation gradient. Their relationship is pretty much like thermal eigenstrain and eigenstrain. So your current design is pretty much like "implementing thermal eigenstrain instead of considering eigenstrain in general".

This is fine, but what if in the future we not only have thermal deformation gradient, but also swelling deformation gradient, stress-induced eigen deformation gradient, etc? It impossible to consider all of them.

The design I suggest is to generalize the implementation of `ComputeDeformationGradient` mimicking the eigenstrain removal in `ComputeSmallStrain`. You can take a look in a class in my app [https://github.com/hugary1995/raccoon/blob/master/src/materials/large_deformation_models/ComputeDeformationGradient.C](https://github.com/hugary1995/raccoon/blob/master/src/materials/large_deformation_models/ComputeDeformationGradient.C) .

Basically, it removes all eigen deformation gradients at once, and leaves you with a mechanical deformation gradient (counterpart of "mechanical strain" in the small deformation context).

**Comment by hugary1995** [comment]
Oh sorry I didn't see @jiangwen84 has already made the same comment. Glad you are already aware of this issue!

**Comment by dewenyushu** [comment]
> Oh sorry I didn't see @jiangwen84 has already made the same comment. Glad you are already aware of this issue!

Thanks @hugary1995 The most recent commit has the initial code to consider eigenstrain in general (see `ComputeCrystalPlasticityEigenstrainBase`). 

In case you are interested, in stead of 'removing' the eigenstrain from the total strain, for CP models, I am introducing additional deformation gradients, like `F = FeFp{Fth...Fi}`, where Fth is the thermal deformation gradient, Fi is the deformation gradient from other physics, for example. The corresponding eigenstrains are computed as Green strains, e_th = (Fth^T*  Fth - I)/2, etc.

File: modules/tensor_mechanics/include/materials/crystal_plasticity/ComputeCrystalPlasticityEigenstrainBase.h [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  MaterialProperty<RankTwoTensor> & _eigenstrain;
+
+  ///Stores the deformation gradient
+  MaterialProperty<RankTwoTensor> & _deformation_gradient;
+  const MaterialProperty<RankTwoTensor> & _deformation_gradient_old;
+
+  /// Substepping time step value used within the inheriting crystal plasticity eigenstrain calculations
+  Real _substep_dt;
+
+  // Not sure about this. Copied from ComputeEigenstrainBase
```
Comment (jiangwen84) [inline modules/tensor_mechanics/include/materials/crystal_plasticity/ComputeCrystalPlasticityEigenstrainBase.h:69]: If we are not going to derive from `ComputeEigenstrainBase` and you think it is not used, let us remove this variable.

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    _temperature(coupledValue("temperature")),
+    _temperature_old(coupledValueOld("temperature")),
+    _ddeformation_gradient_dT(declarePropertyDerivative<RankTwoTensor>(
+        _deformation_gradient_name, getVar("temperature", 0)->name()))
+{
+  std::vector<Real> therm_exp_coeff = getParam<std::vector<Real>>("thermal_expansion_coefficients");
+  if (therm_exp_coeff.size() != _mesh.dimension())
+    paramError("thermal_expansion_coefficients",
+               "thermal expansion coefficient size must match the mesh dimension");
+  _thermal_expansion_coefficients.fillFromInputVector(therm_exp_coeff);
```
Comment (jiangwen84) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C:39]: is `_thermal_expansion_coefficients` always a diagonal matrix and filled by three entries?

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+void
+ComputeCrystalPlasticityThermalEigenstrain::computeQpDeformationGradient()
+{
+  // compute the deformation gradient due to thermal expansion
+  Real dtheta = (_temperature[_qp] - _temperature_old[_qp]) * _substep_dt / _dt;
+  RankTwoTensor residual_equivalent_thermal_expansion_increment =
+      RankTwoTensor::Identity() - dtheta * _thermal_expansion_coefficients;
+  RankTwoTensor _inverse_deformation_gradient =
+      _deformation_gradient_old[_qp].inverse() * residual_equivalent_thermal_expansion_increment;
+  _deformation_gradient[_qp] = _inverse_deformation_gradient.inverse();
```
Comment (jiangwen84) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C:51]: I think we can avoid one inverse operation.

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C [hunk 9]
```diff
@@ -625,3 +688,19 @@ ComputeMultipleCrystalPlasticityStress::lineSearchUpdate(const Real & rnorm_prev
...
+{
+  _inverse_eigenstrain_deformation_grad.zero();
+  _inverse_eigenstrain_deformation_grad.addIa(1.0);
+
+  for (unsigned int i = 0; i < _num_eigenstrains; ++i)
+  {
+    _eigenstrains[i]->setSubstepDt(_substep_dt);
+    _eigenstrains[i]->computeQpProperties();
+    _inverse_eigenstrain_deformation_grad =
+        _inverse_eigenstrain_deformation_grad * _eigenstrains[i]->getDeformationGradientInverse();
```
Comment (jiangwen84) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C:703]: _inverse_eigenstrain_deformation_grad *= _eigenstrains[i]->getDeformationGradientInverse();

File: modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C [hunk 10]
```diff
@@ -329,25 +329,52 @@ CrystalPlasticityStressUpdateBase::calculateSchmidTensor(
...
+    const RankTwoTensor & inverse_eigenstrain_deformation_grad_old,
+    const unsigned int & num_eigenstrains)
 {
   std::vector<Real> dslip_dtau(_number_slip_systems, 0.0);
   std::vector<RankTwoTensor> dtaudpk2(_number_slip_systems);
   std::vector<RankTwoTensor> dfpinvdslip(_number_slip_systems);
 
+  RankTwoTensor eigenstrain_deformation_grad_old =
+      inverse_eigenstrain_deformation_grad_old.inverse();
+
```
Comment (jiangwen84) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C:368]: move those to line 374

**Review by jiangwen84 — CHANGES_REQUESTED** [review]
Overall looks good. I made a few minor comments.

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C (hunk 7 above)
Comment (dewenyushu) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C:39]: Yeah, this is what I've seen so far from the paper.

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C (hunk 8 above)
Comment (dewenyushu) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C:51]: Right, good catch!

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityEigenstrainBase.C [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ComputeCrystalPlasticityEigenstrainBase.h"
+#include "RankTwoTensor.h"
+
+InputParameters
+ComputeCrystalPlasticityEigenstrainBase::validParams()
```
[the remark this answers is not in the document]
Comment (dewenyushu) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityEigenstrainBase.C:14]: Looks reasonable to do so, since the capability and many function names are the same

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C (hunk 7 above)
Comment (jiangwen84) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C:39]: I just took a look at the `fillFromInputVector` method. It can take the vector size of 1, 3, 6 and 9. If your mesh is 2, you probably still need to provide 3 entries, although I am not exactly sure what the third entry will be.

Comment (hugary1995) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C:39]: I don't think it is always diagonal. People usually measure the expansion coefficients `a1`, `a2`, `a3` along three principal axes. The only case this would be diagonal is when the principal axes coincide with the cartesian coords.

Why can't we just get a `RankTwoTensor` material property as the `_thermal_expansion_coefficients`? That way the user has freedom to specify whatever anisotropy in their expansion coefficients.

**Comment by hugary1995** [comment]
I think this is far reaching but not directly related to this PR -- my biggest complaint about the current CP implementation is that, although it is a total formulation, you really only hard-coded in one elastic constitutive relation, which is the Saint Vernant-Kirchhoff one, i.e. S = C E, E = (F^T F - I)/2, and all your plasticity updates assume the same. Although the elasticity model barely makes any difference in the simulations of ductile metals, there are still people who actually care about that. I hope someday you can provide an interface to specify different elastic constitutive laws. Ideally, there should be an interface to define the Mandel stress in the elasticity model.

**Comment by jiangwen84** [comment]
> I think this is far reaching but not directly related to this PR -- my biggest complaint about the current CP implementation is that, although it is a total formulation, you really only hard-coded in one elastic constitutive relation, which is the Saint Vernant-Kirchhoff one, i.e. S = C E, E = (F^T F - I)/2, and all your plasticity updates assume the same. Although the elasticity model barely makes any difference in the simulations of ductile metals, there are still people who actually care about that. I hope someday you can provide an interface to specify different elastic constitutive laws. Ideally, there should be an interface to define the Mandel stress in the elasticity model.

Great point. I think we can wait until we have a good plan on how to implement general multiplicative plasticity formulation.

**Comment by dewenyushu** [comment]
> I think this is far reaching but not directly related to this PR -- my biggest complaint about the current CP implementation is that, although it is a total formulation, you really only hard-coded in one elastic constitutive relation, which is the Saint Vernant-Kirchhoff one, i.e. S = C E, E = (F^T F - I)/2, and all your plasticity updates assume the same. Although the elasticity model barely makes any difference in the simulations of ductile metals, there are still people who actually care about that. I hope someday you can provide an interface to specify different elastic constitutive laws. Ideally, there should be an interface to define the Mandel stress in the elasticity model.

Thanks @hugary1995  and @jiangwen84, really appreciate your inputs here - I will put this in the my note and bring it up when we do planning next time.

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C (hunk 7 above)
Comment (jiangwen84) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C:39]: @hugary1995 I think what @dewenyushu did here is fine. She basically fill in a tensor with a vector. If we do not limit the size of the vector, then users have freedom to specify  whatever tensor they  want.

Comment (dewenyushu) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C:39]: @jiangwen84 Yeah, it totally makes sense now to remove the limit of the size: 1) it does not make sense in 2D, and 2) as @hugary1995 pointed out there is potential need for off diagonal entries in the thermal expansion coefficient

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    _thermal_expansion_coefficients.fillFromInputVector(
+        getParam<std::vector<Real>>("thermal_expansion_coefficients"));
+  }
+  catch (const MooseException & e)
+  {
+    paramError(
+        "thermal_expansion_coefficients",
+        "input vector size should be 1, 3, 6 or 9 to fill the thermal expansion coefficients");
+  }
+}
```
Comment (jiangwen84) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C:47]: I know you spent some time on figuring out how to catch 'MooseException`, but you probably do not need this. When a wrong vector is passed into `fillFromInputVector`, a mooseError will be thrown out to tell users the vector size must be 1, 3, 6 or 9. 

Is my understanding correct? :-)

Comment (dewenyushu) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C:47]: The mooseError from `fillFromInputVector ` does not tell the user directly which line of the input is causing the error. If the user has multiple coefficients like this, it would be hard to debug

Therefore, I suppose catching this exception and throw a parameter error relating to the specific material would be better. 

What do you think?

Comment (jiangwen84) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C:47]: I see your point.  Good!

File: framework/include/utils/RankTwoTensorImplementation.h [hunk 13]
```diff
@@ -146,7 +146,8 @@ void
 RankTwoTensorTempl<T>::fillFromInputVector(const std::vector<T> & input, FillMethod fill_method)
 {
   if (fill_method != autodetect && fill_method != input.size())
-    mooseError("Expected an input vector size of ", fill_method, " to fill the RankTwoTensorTempl");
+    mooseException(
+        "Expected an input vector size of ", fill_method, " to fill the RankTwoTensorTempl");
```
Comment (jiangwen84) [inline framework/include/utils/RankTwoTensorImplementation.h:150]: I am not sure if we should change mooseError to mooseException. For your code, I think you will throw an error at the end, but if others call this function somewhere else, only exception will be generated. Let us bring this issue to our standup meeting tomorrow.

Comment (dewenyushu) [reply framework/include/utils/RankTwoTensorImplementation.h:150]: According to the discussion on slack, it sounds like the `mooseException` would error the same way as `mooseError`, if we do not catch it - Even though it does not seem to break things now, I agree it is better to bring this up for a broader discussion before we move on :)

**Comment by dewenyushu** [comment]
@jiangwen84 this PR should be good to go. Two additional things will be addressed in separate PRs soon:

- Rotation of the thermal expansion coefficient
- Tensor input parameter checking (#17957)

**Comment by jiangwen84** [comment]
> @jiangwen84 this PR should be good to go. Two additional things will be addressed in separate PRs soon:
> 
> * Rotation of the thermal expansion coefficient
> * Tensor input parameter checking (#17957)

@dewenyushu  I think we need to add the thermal expansion tensor rotation in this PR. Otherwise, if this gets merged, users will use the codes but get wrong answers. Sorry for the additional request.

**Comment by dewenyushu** [comment]
> @dewenyushu I think we need to add the thermal expansion tensor rotation in this PR. Otherwise, if this gets merged, users will use the codes but get wrong answers. Sorry for the additional request.

I thought this would take longer. But here is the updated version with rotation considered! :)

**Review by jiangwen84 — CHANGES_REQUESTED** [review]
Looks good. 

One quick question: do you know if there is any way to output the individual eigenstrain and visualize them in Paraview?

**Comment by dewenyushu** [comment]
> Looks good.
> 
> One quick question: do you know if there is any way to output the individual eigenstrain and visualize them in Paraview?

The `eigenstrain` is a material property stored in `ComputeEigenstrainBase`. So I suppose we should be able to output it component-wise via defining corresponding auxvariables and postprocessors.

**Comment by jiangwen84** [comment]
> > Looks good.
> > One quick question: do you know if there is any way to output the individual eigenstrain and visualize them in Paraview?
> 
> The `eigenstrain` is a material property stored in `ComputeEigenstrainBase`. So I suppose we should be able to output it component-wise via defining corresponding auxvariables and postprocessors.

Could you add the output in one of your regression test?

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+{
+}
+
+void
+ComputeCrystalPlasticityThermalEigenstrain::initQpStatefulProperties()
+{
+  ComputeCrystalPlasticityEigenstrainBase::initQpStatefulProperties();
+  // rotate the thermal deforamtion gradient for crystals based on Euler angles
+  _lattice_thermal_expansion_coefficients[_qp] =
+      _thermal_expansion_coefficients.rotated(_crysrot[_qp]);
```
[the remark this answers is not in the document]
Comment (dewenyushu) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C:54]: The `_thermal_expansion_coefficients` is actually a `RankTwoTensor`. During the initialization, I was calling its constructor with the vector as inputs.

[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C:54]: Got it. I missed that.

**Review by jiangwen84 — APPROVED** [review]
